### PR TITLE
Issue #1842, Added ability to download revocation certificate

### PR DESCRIPTION
**Pull request by michael-volynets**
Issue #1842 
Time Spent: 4h

I thought it would be easier :)
I faced some issues that slowed me a little.
1. I had to update typings. First I decided to copy/pasted definitions from here `https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/openpgp/ts3.2/index.d.ts`  but that didn't work so I added that definitions that we only need.
2. I don't know why but `openpgp.key.Key.revoke` returns not `string` but `Stream`. I checked the original method to understand what's going on there. I checked it and didn't find out why so I had to somehow deal with that, and I corrected typings a little, so `openpgp.key.Key.revoke` may return either `string`, `Stream<string>` or `undefined`.
3. And the last thing I had to do is to handle a situation when a user doesn't have stored passphrase. Actually, I didn't have any trouble implementing this step.

### CODE REVIEW

File: extension/chrome/settings/modules/my_key.ts [hunk 1]
```diff
@@ -68,13 +79,35 @@ View.run(class MyKeyView extends View {
     }
   }
 
+  private downloadRevocationCert = async (enteredPP?: string) => {
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key.ts]: I'm thinking that the code could really simplify we used `elements/passphrase.htm` module instead.

Or maybe if we could make Sweetalert prompt where user enters some info, then we could use that for pass phrases which is much simpler.

But for now we can go with this.

File: extension/chrome/settings/modules/my_key.ts [hunk 2]
```diff
@@ -44,20 +46,29 @@ View.run(class MyKeyView extends View {
...
+    await initPassphraseToggle(['input_passphrase']);
   }
 
   setHandlers = () => {
-    $('.action_download_pubkey').click(this.setHandlerPrevent('double', () => this.downloadPubKeyHandler()));
-    $('.action_download_prv').click(this.setHandlerPrevent('double', () => this.downloadPrvKeyHandler()));
-    const clipboardOpts = { text: (trigger: HTMLElement) => trigger.className.includes('action_copy_pubkey') ? this.primaryKi!.public : this.primaryKi!.private };
+    $('.action_download_revocation_cert').click(this.setHandlerPrevent('double', () => this.downloadRevocationCert()));
+    $('.action_continue_download').click(this.setHandlerPrevent('double', () => this.downloadRevocationCert(String($('#input_passphrase').val()))));
+    $('#input_passphrase').on('keydown', this.setHandler((el, ev) => {
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key.ts]: Can use `this.setEnterHandlerThatClicks`

File: extension/chrome/settings/modules/my_key.ts [hunk 3]
```diff
@@ -44,20 +46,29 @@ View.run(class MyKeyView extends View {
...
+      if (ev.which === 13) {
+        $('.action_continue_download').click();
+      }
+    }));
+    $('.action_cancel_download_cert').click(this.setHandler(() => { $('.enter_pp').hide(); }));
+    $('.action_download_pubkey').click(this.setHandlerPrevent('double', this.downloadPubKeyHandler));
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key.ts]: should use `() => this...` for handlers, for consistency.

File: extension/js/common/core/types/openpgp.d.ts [hunk 4]
```diff
@@ -850,6 +875,11 @@ declare namespace OpenPGP {
       getKeyId(): Keyid;
     }
 
+    export interface revoke_reasonForRevocation {
```
Comment (tomholub) [inline extension/js/common/core/types/openpgp.d.ts]: No need to export it

File: extension/js/common/core/types/openpgp.d.ts [hunk 5]
```diff
@@ -818,6 +841,8 @@ declare namespace OpenPGP {
       update(key: Key): void;
       verifyPrimaryKey(): Promise<enums.keyStatus>;
       isRevoked(): Promise<boolean>;
+      revoke(easonForRevocation: revoke_reasonForRevocation, date?: Date): Promise<Key>;
```
Comment (tomholub) [inline extension/js/common/core/types/openpgp.d.ts]: typo `eason`

File: extension/js/common/core/pgp-key.ts [hunk 6]
```diff
@@ -354,4 +354,15 @@ export class PgpKey {
     throw new Error('No valid signature found in key');
   }
 
+  static revoke = async (key: OpenPGP.key.Key) => {
```
Comment (tomholub) [inline extension/js/common/core/pgp-key.ts]: Will add signature for clarity

File: extension/chrome/settings/modules/my_key.ts [hunk 7]
```diff
@@ -68,13 +79,35 @@ View.run(class MyKeyView extends View {
...
+          return;
+        }
+      } else {
+        $('.enter_pp').show();
+        return;
+      }
+    }
+    $('.enter_pp').hide();
+    $('#input_passphrase').val('');
+    const revokedArmored = await PgpKey.revoke(prv);
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key.ts]: Should test if not undefined, and alert instead

File: extension/js/common/core/types/openpgp.d.ts [hunk 8]
```diff
@@ -704,6 +704,29 @@ declare namespace OpenPGP {
       signature = 6,
     }
 
+    enum reasonForRevocation {
+      /**
```
Comment (tomholub) [inline extension/js/common/core/types/openpgp.d.ts]: will compact the comments a bit

File: extension/js/common/core/types/openpgp.d.ts [hunk 9]
```diff
@@ -704,6 +704,29 @@ declare namespace OpenPGP {
...
+       * No reason specified (key revocations or cert revocations)
+       */
+      no_reason = 0,
+      /**
+       * Key is superseded (key revocations)
+       */
+      key_superseded = 1,
+      /**
+       * Key material has been comPromise<any>d (key revocations)
```
Comment (tomholub) [inline extension/js/common/core/types/openpgp.d.ts]: typo

**Comment by tomholub** [comment]
Todo - still investigate output format (maybe should say revocation)

**Comment by michael-volynets** [comment]
Thanks!

**Pull request merged**